[BEAM-4711] fix globbing in LocalFileSystem.delete

ryan-williams opened the pull request:
Make `LocalFileSystem.delete` support globs, instead of just literal paths

This makes its behavior consistent with e.g. `GCSFileSystem`

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

ryan-williams commented:
I never pinged for review here; @lukecwik do you care to take a look?

lukecwik commented:
Sorry about the delay, was on vacation.

lukecwik commented on sdks/python/apache_beam/io/localfilesystem.py:
```diff
@@ -295,11 +296,20 @@ def _delete_path(path):
+      if not expanded:
+        try_delete(path)
+      else:
+        for p in expanded:
```
I believe we should be relying on `filesystem.match` for globbing to have a consistent *globbing* experience across the filesystem APIs. This would be similar to what we do within the [gcsfilesystem.py](https://github.com/apache/beam/blob/1348cae2daaa403867fba192e4e06ff4addbf5ae/sdks/python/apache_beam/io/gcp/gcsfilesystem.py#L313)

ryan-williams replied:
good idea, done

ryan-williams commented:
I addressed your comment and added some tests, thanks @lukecwik 

The tests are a bit more involved than "simple" coverage you asked for; better safe than sorry when dealing with filesystem-deletion code paths, I figure…

lukecwik commented on sdks/python/apache_beam/io/localfilesystem.py:
```diff
@@ -321,11 +321,22 @@ def _delete_path(path):
+    for match_result in self.match(paths):
+      metadata_list = match_result.metadata_list
+
+      if not metadata_list:
```
Why do you want to make this an error? (does another filesystem do this or a comment say this is the right contract?)

I would suspect that empty matches turn into no-ops.

lukecwik replied:
Turns out that we test that delete throws an error if non existent and that we have an inconsistent definition between multiple implementations (e.g Java LocalFileSystem). Filed https://issues.apache.org/jira/browse/BEAM-5425 to ask whether the contract should be changed to allow prior partial successes to continue and for errors to not be thrown for non-existent files.

ryan-williams replied:
sorry I dropped this; yea I only made it an error to conform to the existing test you mentioned

lukecwik commented:
Run Python PreCommit
